File: ch2/04_stack_queue.java
import java.util.*;

class ch2_04_stack_queue {
  public static void main(String[] args) {
    Stack<Character> s = new Stack<Character>();

    // Queue is abstract, must be instantiated with LinkedList
    // (special case for Java Queue)
    Queue<Character> q = new LinkedList<Character>();

    Deque<Character> d = new LinkedList<Character>();

    System.out.println(s.isEmpty());          // currently s is empty, true
    System.out.println("==================");
    s.push('a');
    s.push('b');
    s.push('c');
    // stack is LIFO, thus the content of s is currently like this:
    // c <- top
    // b
    // a
    System.out.println(s.peek());                             // output 'c'
    s.pop();                                                 // pop topmost
    System.out.println(s.peek());                             // output 'b'
    System.out.println(s.empty());       // currently s is not empty, false
    System.out.println("==================");

    System.out.println(q.isEmpty());          // currently q is empty, true
    System.out.println("==================");
    while (!s.isEmpty()) {                // stack s still has 2 more items
      q.offer(s.peek()); // enqueue 'b', and then 'a' (the method name in Java Queue for push/enqueue operation is 'offer')
      s.pop();
    }
    q.offer('z');                                      // add one more item
    System.out.println(q.peek());                             // prints 'b'
    // in Java, it is harder to see the back of the queue...

    // output 'b', 'a', then 'z' (until queue is empty), according to the insertion order above
    System.out.println("==================");
    while (!q.isEmpty()) {
      System.out.printf("%c\n", q.peek());         // take the front first
      q.poll();                         // before popping (dequeue-ing) it
    }

    System.out.println("==================");
    d.addLast('a');
    d.addLast('b');
    d.addLast('c');
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'a - c'
    d.addFirst('d');
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'd - c'
    d.pollLast();
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'd - b'
    d.pollFirst();
    System.out.printf("%c - %c\n", d.getFirst(), d.getLast()); // prints 'a - b'
  }
}

File: ch2/04_stack_queue.ml
open Printf

let int_of_bool b = if b then 1 else 0

let () =
  let s = Stack.create () in
  let q = Queue.create () in

  printf "%d\n" (int_of_bool (Stack.is_empty s));
  printf "==================\n";
  Stack.push 'a' s;
  Stack.push 'b' s;
  Stack.push 'c' s;

  printf "%c\n" (Stack.top s);
  Stack.pop s |> ignore;
  printf "%c\n" (Stack.top s);
  printf "%d\n" (int_of_bool (Stack.is_empty s));
  printf "==================\n";

  printf "%d\n" (int_of_bool (Queue.is_empty q));
  printf "==================\n";
  while (not (Stack.is_empty s)) do
    Queue.push (Stack.pop s) q;
  done;
  Queue.push 'z' q;
  printf "%c\n" (Queue.top q);

  printf "==================\n";
  while (not (Queue.is_empty q)) do
    printf "%c\n" (Queue.pop q);
  done;
